docs: add request storage guide (#521)
apify/crawlee-python: docs/guides/request_storage.mdx

---
id: request-storage
title: Request storage
description: How to store the requests your crawler will go through
---

import ApiLink from '@site/src/components/ApiLink';

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import CodeBlock from '@theme/CodeBlock';

import RqBasicExample from '!!raw-loader!./code/request_storage_rq_basic.py';
import RqWithCrawlerExample from '!!raw-loader!./code/request_storage_rq_with_crawler.py';
import RqWithCrawlerExplicitExample from '!!raw-loader!./code/request_storage_rq_with_crawler_explicit.py';

import RlBasicExample from '!!raw-loader!./code/request_storage_rl_basic.py';
import RlWithCrawlerExample from '!!raw-loader!./code/request_storage_rl_with_crawler.py';

import RsDoNotPurgeExample from '!!raw-loader!./code/request_storage_do_not_purge.py';
import RsPurgeExplicitlyExample from '!!raw-loader!./code/request_storage_purge_explicitly.py';

This guide explains the different types of request storage available in Crawlee, how to store the requests that your crawler will process, and which storage type to choose based on your needs.

## Request providers overview

> All request storage types in Crawlee implement the same interface - <ApiLink to="class/RequestProvider">`RequestProvider`</ApiLink>. This unified interface allows them to be used in a consistent manner, regardless of the storage backend. The request providers are managed by storage clients - subclasses of <ApiLink to="class/BaseStorageClient">`BaseStorageClient`</ApiLink>. For instance, <ApiLink to="class/MemoryStorageClient">`MemoryStorageClient`</ApiLink> stores data in memory while also can offload them to the local directory. Data are stored in the following directory structure:

Reviewer: ```suggestion
All request storage types in Crawlee implement the same interface - <ApiLink to="class/RequestProvider">`RequestProvider`</ApiLink>. This unified interface allows them to be used in a consistent manner, regardless of the storage backend. The request providers are managed by storage clients - subclasses of <ApiLink to="class/BaseStorageClient">`BaseStorageClient`</ApiLink>. For instance, <ApiLink to="class/MemoryStorageClient">`MemoryStorageClient`</ApiLink> stores data in memory while it can also offload them to the local directory. Data are stored in the following directory structure:
```

implemented. It is currently intended mainly for testing purposes and small-scale projects. The current
implementation is only in-memory storage and is very limited. It will be (re)implemented in the future.
For more details, see the GitHub issue [crawlee-python#99](https://github.com/apify/crawlee-python/issues/99).
For production usage we recommend to use the <ApiLink to="class/RequestQueue">`RequestQueue`</ApiLink>.

:::

The following code demonstrates the usage of the <ApiLink to="class/RequestList">`RequestList`</ApiLink>:

<Tabs groupId="request_list">
    <TabItem value="request_list_basic_example" label="Basic usage" default>
        <CodeBlock className="language-python">
            {RlBasicExample}
        </CodeBlock>
    </TabItem>
    <TabItem value="request_list_with_crawler" label="Usage with Crawler">
        <CodeBlock className="language-python">
            {RlWithCrawlerExample}
        </CodeBlock>
    </TabItem>
</Tabs>

{/*

## Which one to choose?

TODO: write this section, once https://github.com/apify/crawlee-python/issues/99 is resolved

*/}

## Cleaning up the storages

> By default, storages are automatically purged before the crawler starts, unless explicitly configured otherwise, for that case, see <ApiLink to="class/Configuration#purge_on_start">`Configuration.purge_on_start`</ApiLink>. This cleanup happens as soon as a storage is accessed, either when you open a storage (e.g., using <ApiLink to="class/RequestQueue#open">`RequestQueue.open`</ApiLink>) or when interacting with a storage through one of the helper methods (e.g., <ApiLink to="class/BasicCrawler#add_requests">`BasicCrawler.add_requests`</ApiLink>, which implicitly opens the request storage).

Reviewer: Default storages, that is :slightly_smiling_face:

Author: Is it true? It opens the configured storage, does not have to be the default one.

Reviewer 2: only default storages are auto-purged

implemented. It is currently intended mainly for testing purposes and small-scale projects. The current
implementation is only in-memory storage and is very limited. It will be (re)implemented in the future.
For more details, see the GitHub issue [crawlee-python#99](https://github.com/apify/crawlee-python/issues/99).
For production usage we recommend to use the <ApiLink to="class/RequestQueue">`RequestQueue`</ApiLink>.

:::

The following code demonstrates the usage of the <ApiLink to="class/RequestList">`RequestList`</ApiLink>:

<Tabs groupId="request_list">
    <TabItem value="request_list_basic_example" label="Basic usage" default>
        <CodeBlock className="language-python">
            {RlBasicExample}
        </CodeBlock>
    </TabItem>
    <TabItem value="request_list_with_crawler" label="Usage with Crawler">
        <CodeBlock className="language-python">
            {RlWithCrawlerExample}
        </CodeBlock>
    </TabItem>
</Tabs>

{/*

## Which one to choose?

TODO: write this section, once https://github.com/apify/crawlee-python/issues/99 is resolved

*/}

## Request-related helpers

We offer several helper functions to simplify interactions with request storages:

> - The <ApiLink to="class/AddRequestsFunction">`add_requests`</ApiLink> function allows you to manually add specific URLs to the configured request storage. In this case, you must explicitly provide the URLs you want to be added to the request storage.

Reviewer: ```suggestion
- The <ApiLink to="class/AddRequestsFunction">`add_requests`</ApiLink> function allows you to manually add specific URLs to the configured request storage. In this case, you must explicitly provide the URLs you want to be added to the request storage. If you need to specify further details of the request, such as a `label` or `user_data`, you have to pass instances of the <ApiLink to="class/Request">`Request`</ApiLink> class to the helper.
```

> - The <ApiLink to="class/EnqueueLinksFunction">`enqueue_links`</ApiLink> function is designed to find new URLs in the current page and add them to the request storage. You can specify different enqueue strategies or add excluding / including filters to limit the URLs that will be added to the request storage. See [Crawl website with relative links](../examples/crawl-website-with-relative-links) for more details.

Reviewer: Maybe we could mention that it works fine without arguments... and we should probably mention selectors.

Author: There is already an example that demonstrates it works without arguments. Also, in my opinion, we should not describe the function interface & arguments in the guides (too much). For such cases should serve the API docs. It would be redundant and also hard to maintain.

Reviewer: I mostly agree, it just feels weird to mention strategies and omit selectors...

Author: Updated

> - The <ApiLink to="class/EnqueueLinksFunction">`enqueue_links`</ApiLink> function is designed to discover new URLs in the current page and add them to the request storage. It can be used with default settings, requiring no arguments, or you can customize its behavior by specifying URL selectors, choosing different enqueue strategies, or applying include/exclude filters to control which URLs are added. See [Crawl website with relative links](../examples/crawl-website-with-relative-links) example for more details.

Reviewer: ```suggestion
- The <ApiLink to="class/EnqueueLinksFunction">`enqueue_links`</ApiLink> function is designed to discover new URLs in the current page and add them to the request storage. It can be used with default settings, requiring no arguments, or you can customize its behavior by specifying link element selectors, choosing different enqueue strategies, or applying include/exclude filters to control which URLs are added. See [Crawl website with relative links](../examples/crawl-website-with-relative-links) example for more details.
```
